src/App.tsx

```diff
@@ -1,26 +1,301 @@
+  const [updatingIds, setUpdatingIds] = useState<number[]>([]);
+  const [editingTodoId, setEditingTodoId] = useState<number | null>(null);
+  const [updateErrorId, setUpdateErrorId] = useState<number | null>(null);
+
+  useEffect(() => {
+    if (!USER_ID) {
+      return;
+    }
+
+    setIsLoading(true);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage(ErrorMessage.LoadTodos))
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (!isAdding && loadingIds.length === 0) {
+      inputRef.current?.focus();
+    }
+  }, [isAdding, loadingIds]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const handleAddTodo = (title: string) => {
+    const trimmedTitle = title.trim();
+
+    if (!trimmedTitle) {
+      setErrorMessage(ErrorMessage.EmptyTitle);
+
+      return Promise.resolve();
+    }
+
+    setIsAdding(true);
+
+    const newTodo = {
+      id: 0,
+      title: trimmedTitle,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo(newTodo);
+
+    return addTodo(newTodo)
+      .then(createdTodo => {
+        setTodos(prev => [...prev, createdTodo]);
+        setNewTodoTitle('');
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.AddTodo);
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsAdding(false);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingIds(prev => [...prev, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.DeleteTodo);
+      })
+      .finally(() => {
+        setLoadingIds(prev => prev.filter(id => id !== todoId));
+      });
+  };
+
+  const handleStartEditing = (id: number) => {
+    setEditingTodoId(id);
+    setUpdateErrorId(null);
+  };
+
+  const handleCancelEditing = () => {
+    setEditingTodoId(null);
+    setUpdateErrorId(null);
+  };
+
+  const handleUpdateTodo = async (todoId: number, newTitle?: string) => {
+    const todo = todos.find(t => t.id === todoId);
+
+    if (!todo) {
+      return;
+    }
+
+    if (newTitle !== undefined && newTitle.trim() === '') {
+      setUpdatingIds(prev => [...prev, todoId]);
+      setErrorMessage(null);
+
+      try {
+        await deleteTodo(todoId);
+        setTodos(prev => prev.filter(t => t.id !== todoId));
+        setEditingTodoId(null);
+        setUpdateErrorId(null);
+      } catch (error) {
+        setErrorMessage(ErrorMessage.DeleteTodo);
+        setUpdateErrorId(todoId);
+      } finally {
+        setUpdatingIds(prev => prev.filter(id => id !== todoId));
+      }
+
+      return;
+    }
+
+    const updatedFields: Partial<Todo> = {
+      completed: newTitle === undefined ? !todo.completed : todo.completed,
+    };
+
+    if (newTitle !== undefined) {
```

Comment:
```suggestion
    if (!newTitle) {
```
looks bad use clean check